campaignion_tracking/package/src/gtm.js

```diff
@@ -154,11 +171,59 @@ export class GTMTracker {
     return gtmData
   }
 
-  updateContext (context) {
-    Object.assign(this._context.donation, context.node, context.donation, context.webform)
+  /**
+   * Maintains a context between event handling.
+   *
+   * You can provide context with a first event and call this method during
+   * handling.
+   * When another event arrives it can also enrich this context and read former
+   * context data as well.
+   *
+   * This context is saved in the browser storage to persist between page
+   * loads. It is reset on donationSuccess.
+   *
+   * @param {context} context Tracking context data.
+   */
+  updateContext (context = {}) {
+    // Check if context exists
+    if (context.donation) {
+      Object.assign(this._context.donation, context.donation)
+    }
+    if (context.node) {
+      Object.assign(this._context.node, context.node)
+    }
+    if (context.webform) {
+      Object.assign(this._context.webform, context.webform)
+    }
     this.saveToStorage('context', this._context)
   }
 
+  /**
+   * Handle "submission".
+   *
+   * Event data: { nid, sid, title }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_submission (eventName, eventData, context) { // eslint-disable-line camelcase
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+
+    let data = {
+      event: 'submission',
+      webform: {
+        nid: eventData.nid || null,
+        sid: eventData.sid || null,
+        title: eventData.title || null,
+      }
+    }
+    // Allow others to modify the data being sent to GTM.
+    data = this.callChangeHook(eventName, data, this._context)
```

Comment:
Is there a reason to implement this in hook-style as opposed to JS events?

Reply:
how would you receive back the modified data then? :thinking: 

but I agree, there might be opportunity for refactoring here...